File: algoexpert/non_constructible_change.py
def all_same(arr):
	return len(set(arr)) == 1

def is_sumable(arr, target):
	data = {}
	
	for i, v in enumerate(arr):
		if data.get(v) is None:
			data[target - v] = i
		else:
			return True
			
	return False
	
def nonConstructibleChange(coins):
    # Write your code here.
	if (len(coins) == 0):
		return 1
	
	if all_same(coins):
		return (len(coins) * coins[0]) + 1
	
	coins.sort()
	
	if coins[0] != 1:
		return 1

	for target in range(1, coins[len(coins) - 1]):
		if target == 1:
			continue

		if is_sumable(coins, target) == False:
			return target


print(nonConstructibleChange([5, 7, 1, 1, 2, 3, 22]))

File: algoexpert/non_constructible_change.go
package main

import (
	"fmt"
	"sort"
)

func uniqueArr(arr []int) []int {
	var uniqueArr []int
	var data map[int]bool

	for _, a := range arr {
		if _, ok := data[a]; !ok {
			data[a] = true
			uniqueArr = append(uniqueArr, a)
		}
	}

	return uniqueArr
}

func NonConstructibleChange(coins []int) int {
	if len(coins) == 0 {
		return 1
	}

	uniqueCoins := uniqueArr(coins)

	if len(uniqueCoins) == 1 {
		return uniqueCoins[0] + 1
	}

	sort.Ints(uniqueCoins)

	if uniqueCoins[0] == 1 {
		return 1
	}

	lastCoin := uniqueCoins[len(coins)-1]

	for i := 2; i <= lastCoin; i++ {

	}

	return 0
}

func main() {
	// 1 2 3 5 7 22
	fmt.Println(NonConstructibleChange([]int{5, 7, 1, 1, 2, 3, 22}))
}
